### PR TITLE
Feat/support pipeline parallel

### PR DESCRIPTION
**Pull request by huangting4201**
## Motivation

Support pipeline parallel training.

## Modification

Some important updates:
- `internlm/core/scheduler/pipeline_scheduler.py`: new add dir `internlm/core/scheduler/`, and define pipeline scheduler for pipeline parallel scheduling
- `internlm/initialize/initialize_trainer.py`: modify func `initialize_trainer()` for supporting pipeline parallel 

## Use cases (Optional)

User can use pipeline parallel training by setting parallel config in `configs/7B_sft.py` file:
```python
parallel = dict(
    zero1=8,
    pipeline=2,  # pipeline parallel size, defaults to 1
)
```

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: internlm/initialize/initialize_trainer.py [hunk 1]
```diff
@@ -53,12 +59,16 @@ def initialize_trainer(
     elif isinstance(model, Callable):
         model = model().to(get_current_device())
 
+    gpc.config.grad_scaler.pop("fp16")
```
Comment (sunpengsdu) [inline internlm/initialize/initialize_trainer.py:62]: 这个pop我们还需要吗？

File: internlm/initialize/initialize_trainer.py [hunk 2]
```diff
@@ -67,8 +77,36 @@ def initialize_trainer(
...
+                scatter_gather_tensors=scatter_gather,
+            )
+        else:
+            scheduler = PipelineScheduler(
+                gpc.config.NUM_MICRO_BATCHES, tensor_shape=tensor_shape, scatter_gather_tensors=scatter_gather
+            )
+    else:
+        scheduler = NonPipelineScheduler(gradient_accumulation_size=gpc.config.data.gradient_accumulation)
+
+    # if bf16 is used, this value will be wrongly set to fp32, so it needs to be corrected manually
```
Comment (sunpengsdu) [inline internlm/initialize/initialize_trainer.py:105]: 这些hardcode的逻辑是否还有必要存在

Comment (huangting4201) [reply internlm/initialize/initialize_trainer.py:105]: 之前这个是因为啥会被wrongly set to fp32来着？这块代码我记得我们当时有记录说，后面加pp的话要加回来

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,881 @@
...
+                    data_dict["label"] = element[offset : offset + self.microbatch_size]
+            if data_dict:
+                return data_dict
+            return [val[offset : offset + self.microbatch_size] for val in data]
+        elif isinstance(data, dict):
+            return {k: v[offset : offset + self.microbatch_size] for k, v in data.items()}
+        else:
+            raise TypeError(f"Expected data to be of type torch.Tensor, list, tuple, or dict, but got {type(data)}")
+
+    def load_micro_batch(self):
```
Comment (sunpengsdu) [inline internlm/core/scheduler/pipeline_scheduler.py:186]: 这块后面是不是改下，和no_pp的scheduler共享相同的mciro_batch_data处理的逻辑？ @mwiacx

File: internlm/initialize/launch.py [hunk 4]
```diff
@@ -140,6 +140,10 @@ def args_sanity_check():
     torch.backends.cudnn.deterministic = gpc.config.get("cudnn_deterministic", False)
     clip_grad_norm = gpc.config.hybrid_zero_optimizer.get("clip_grad_norm", 0.0)
 
+    gpc.config.hybrid_zero_optimizer.zero_overlap_communication = True
+    if gpc.config.parallel.pipeline > 1:
+        gpc.config.hybrid_zero_optimizer.zero_overlap_communication = False
```
[the remark this answers is not in the document]
Comment (huangting4201) [reply internlm/initialize/launch.py:145]: 这个应该是放在下一个pr里，支持pp overlap_communication时做

File: internlm/core/scheduler/pipeline_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,881 @@
...
+                "which is the tensors passed by the previous pipeline stage and the dataloader output from this stage, "
+                "i.e. data_process_func(stage_output, dataloader_output)."
+            )
+
+        super().__init__(data_process_func=data_process_func)
+
+        assert num_microbatches > 0, f"expected num_microbatches to be larger then 1, but got {num_microbatches}"
+
+        self.num_microbatches = num_microbatches
+        self.dtype = torch.float
```
Comment (sunpengsdu) [inline internlm/core/scheduler/pipeline_scheduler.py:134]: 大概是这里改的

File: internlm/initialize/launch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunpengsdu) [reply internlm/initialize/launch.py:145]: ˙之前说的overlap卡是因为optimizer那里默认self.skip_grad_reduce = False，也就是啥都不管的情况下，默认全部microbatch进行reduce操作，所以在pp下就卡了。感觉更合理的方式是这个PR里按照no_pp scheduler做法把optimizer传进去，然后里面控制skip_grad_reduce统一设为True，然后后面PR再看如何控制调度让skip_grad_reduce管理通信调度overlap @huangting4201 @SolenoidWGT

File: internlm/initialize/initialize_trainer.py (hunk 1 above)
Comment (huangting4201) [reply internlm/initialize/initialize_trainer.py]: 不需要了，去掉了

File: internlm/initialize/initialize_trainer.py (hunk 2 above)
Comment (huangting4201) [reply internlm/initialize/initialize_trainer.py]: 去掉了，改成在scheduler初始化时传入了dtype

File: internlm/core/scheduler/pipeline_scheduler.py (hunk 3 above)
Comment (huangting4201) [reply internlm/core/scheduler/pipeline_scheduler.py:145]: 这块按照和鹏哥周五的讨论改好了，鹏哥再帮忙review看看

File: internlm/core/scheduler/pipeline_scheduler.py (hunk 5 above)
Comment (huangting4201) [reply internlm/core/scheduler/pipeline_scheduler.py]: 嗯嗯是的

File: internlm/initialize/launch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (huangting4201) [reply internlm/initialize/launch.py]: 新增pp scheduler支持 optimizer overlap communication in 80cbb75f0efbc7f5c79310c3401b729ea173f686

File: internlm/core/communication/utils.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    end_index = start_index + partition_size
+    if new_buffer:
+        data = torch.empty(partition_size, dtype=tensor.dtype, device=torch.cuda.current_device(), requires_grad=False)
+        data.copy_(tensor.view(-1)[start_index:end_index])
+    else:
+        data = tensor.view(-1)[start_index:end_index]
+    return data
+
+
+def gather_split_1d_tensor(tensor: torch.Tensor) -> torch.Tensor:
```
Comment (sunpengsdu) [inline internlm/core/communication/utils.py:115]: 这个函数似乎没有用了

File: internlm/core/communication/utils.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        else:
+            obj_shape = []
+            for _ in range(recv_obj_nums.item()):
+                recv_shape = recv_meta_helper(prev_rank, tensor_kwargs)
+                obj_shape.append(torch.Size(recv_shape))
+
+    return obj_shape
+
+
+def split_tensor_into_1d_equal_chunks(tensor: torch.Tensor, new_buffer=False) -> torch.Tensor:
```
Comment (sunpengsdu) [inline internlm/core/communication/utils.py:94]: 这个似乎也没用到？

File: internlm/utils/model_checkpoint.py [hunk 8]
```diff
@@ -138,11 +138,13 @@ def save_optimizer_checkpoint(optim, state_path):
...
     tp_rank = gpc.get_local_rank(ParallelMode.TENSOR)
     pp_rank = gpc.get_local_rank(ParallelMode.PIPELINE)
+    tp_size = gpc.get_world_size(ParallelMode.TENSOR)
+    pp_size = gpc.get_world_size(ParallelMode.PIPELINE)
     fp = f"optimizer_tp{tp_rank}_pp{pp_rank}_zo{zero_rank}.pt"
 
     states = optim.state_dict()
     if isinstance(optim, HybridZeroOptimizer):
-        if gpc.get_global_rank() < optim.zero_world_size:
+        if gpc.get_global_rank() < optim.zero_world_size * tp_size * pp_size:
```
Comment (sunpengsdu) [inline internlm/utils/model_checkpoint.py:147]: 为啥要新增这个判断？

File: internlm/core/trainer.py [hunk 9]
```diff
@@ -113,7 +118,7 @@ def __init__(
             self._schedule = schedule
 
         if self.uses_pipeline:
-            self._schedule.pre_processing(self)
+            self._schedule.pre_processing(self._engine)
```
Comment (sunpengsdu) [inline internlm/core/trainer.py:121]: 感觉可以不用        if self.uses_pipeline: 来判断，都执行 self._schedule.pre_processing(self._engine)，pre_processing是base类定义的

File: internlm/initialize/initialize_trainer.py [hunk 10]
```diff
@@ -67,8 +75,36 @@ def initialize_trainer(
             handler = PipelineSharedModuleGradientHandler(model=model, optimizer=optimizer)
             gradient_handlers.append(handler)
 
-    scheduler = NonPipelineScheduler(gradient_accumulation_size=gpc.config.data.gradient_accumulation)
-
+    # initialize scheduler for trainer
+    scheduler = None
+    if gpc.is_using_pp():
+        gpc.config.NUM_MICRO_BATCHES = gpc.config.data.micro_num
+        tensor_shape = get_tensor_shape()
```
Comment (sunpengsdu) [inline internlm/initialize/initialize_trainer.py:82]: @yingtongxiong 这里你需要注意下，似乎和padding那里有一些冲突

Comment (sunpengsdu) [reply internlm/initialize/initialize_trainer.py:82]: 后面 @yingtongxiong 会修订

File: internlm/core/trainer.py (hunk 9 above)
Comment (huangting4201) [reply internlm/core/trainer.py]: 已更新

File: internlm/utils/model_checkpoint.py (hunk 8 above)
Comment (huangting4201) [reply internlm/utils/model_checkpoint.py:147]: 修复ckpt的bug
